### PR TITLE
Added seed splitting and recovery to multiwallet

### PR DESCRIPTION
**Pull request by jimmysong**
You can now split and recover your seed using multiwallet!

The commands are `split_seed` and `recover_seed` respectively.

Currently only supports 12 and 24 word seeds.

### CODE REVIEW

File: multiwallet.py [hunk 1]
```diff
@@ -465,6 +468,56 @@ def do_generate_seed(self, arg):
...
+        while True:
+            share_phrase = input(
+                blue_fg("Enter a SLIP39 share (blank to end): ")
+            ).strip()
+            if share_phrase == "":
+                break
+            share_mnemonics.append(share_phrase)
+        passphrase = b""
+        if self.ADVANCED_MODE:
+            has_passphrase = _get_bool("Is there a passphrase?")
```
Comment (mflaxman) [inline multiwallet.py:483]: Set default as N

File: multiwallet.py [hunk 2]
```diff
@@ -465,6 +468,56 @@ def do_generate_seed(self, arg):
...
+        )
+        n = _get_int(
+            "How many shares do you want to generate?",
+            default=2 * k - 1,
+            minimum=k,
+            maximum=199,
+        )
+        passphrase = b""
+        if self.ADVANCED_MODE:
+            add_passphrase = _get_bool("Do you want to add a passphrase?")
```
Comment (mflaxman) [inline multiwallet.py:509]: Add default (false)

File: multiwallet.py [hunk 3]
```diff
@@ -465,6 +468,56 @@ def do_generate_seed(self, arg):
...
+            if add_passphrase:
+                passphrase = _get_confirmed_pw().encode("ascii")
+        shares = ShareSet.generate_shares(mnemonic, k, n, passphrase=passphrase)
+        share_mnemonics = "\n\n".join(shares)
+        if passphrase:
+            additional = " and your passphrase"
```
Comment (mflaxman) [inline multiwallet.py:515]: Make and CAPS

**Review by mflaxman — APPROVED** [review]
Nice!

Besides nits, my only substantive comment is that before giving out shares the script should test every combination to confirm they will recover the original secret.

Perhaps that's overkill, but I think the CPU cost is well worth it.

File: multiwallet.py [hunk 4]
```diff
@@ -465,6 +468,56 @@ def do_generate_seed(self, arg):
             ),
         )
 
+    def do_recover_seed(self, arg):
+        """Recover a seed from shares per SLIP39"""
```
Comment (mflaxman) [inline multiwallet.py:472]: I think this should use the phrase Shamir somewhere (in description and/or do_foo method name) as that's what people know this as. There are (terrible) splitting/recovering algorithms that literally cut the phrase into parts (including Ian Coleman's BIP39 which is surprising and unfortunate).

File: multiwallet.py [hunk 5]
```diff
@@ -465,6 +468,56 @@ def do_generate_seed(self, arg):
...
+            if has_passphrase:
+                passphrase = _get_confirmed_pw().encode("ascii")
+        try:
+            mnemonic = ShareSet.recover_mnemonic(share_mnemonics, passphrase=passphrase)
+            print_green(f"Here is your recovered mnemonic:\n\n{mnemonic}")
+        except (TypeError, ValueError, SyntaxError) as e:
+            print_red(e)
+
+    def do_split_seed(self, arg):
+        """Split a seed to shares per SLIP39"""
```
Comment (mflaxman) [inline multiwallet.py:493]: Ditto on Shamir in name

**Review by mflaxman — APPROVED** [review]
More comments

**Comment by jimmysong** [comment]
Testing every combination when you have like 21-of-40 is a lot (100 billion). I'll cap it at 1000 different combos, which can do 6-of-12.

**Comment by mflaxman** [comment]
You could also ignore order, then the # of permutations is low.

The idea is a basic test to prove it can work, mostly as an extra layer of defensive programming.

**Comment by jimmysong** [comment]
Some really big numbers revealed some stuff I missed in the spec. The most combinations is 8-of-16 which is 12870 combinations. I limited it to 1000 combinations, which took some time, but not too long on my computer.

**Pull request merged**